core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -121,6 +121,33 @@ class KafkaApis(val requestChannel: RequestChannel,
   val adminZkClient = new AdminZkClient(zkClient)
   private val alterAclsPurgatory = new DelayedFuturePurgatory(purgatoryName = "AlterAcls", brokerId = config.brokerId)
 
+  /**
+   * The template to create a forward request handler.
+   */
+  private[server] abstract class ForwardRequestHandler(request: RequestChannel.Request, isForwardRequest: Boolean,
```

Comment:
It seems like we're trying to reuse this handler from the previous patch, but I'm not sure it still makes as much sense. A simpler structure might be something like the following:

```scala
  private def maybeForward(
    request: RequestChannel.Request,
    handler: RequestChannel.Request => Unit
  ): Unit = {
    if (!controller.isActive && config.redirectionEnabled && request.context.principalSerde.isPresent) {
      redirectionManager.forwardRequest(sendResponseMaybeThrottle, request)
    } else {
      // When IBP is smaller than 2.8 or the principal serde is undefined, forwarding is not supported,
      // therefore requests are handled directly.
      handler(request)
    }
  }

  // then invoked like this
override def handle(request: RequestChannel.Request): Unit = {
    try {
      trace(s"Handling request:${request.requestDesc(true)} from connection ${request.context.connectionId};" +
        s"securityProtocol:${request.context.securityProtocol},principal:${request.context.principal}")

      request.header.apiKey match {
      ...
        case ApiKeys.ALTER_CONFIGS => maybeForward(request, handleAlterConfigsRequest)
...


  // unchanged
  def handleAlterConfigs(request): Unit 
```